Mathlib/NumberTheory/LSeries/Dirichlet.lean

```diff
@@ -0,0 +1,415 @@
+    exact LSeriesSummable_mul χ <| ArithmeticFunction.LSeriesSummable_moebius_iff.mpr hs
+
+
+/-!
+### L-series of Dirichlet characters do not vanish on re s > 1
+-/
+
+/-- The L-series of a Dirichlet character does not vanish on the right half-plane `re s > 1`. -/
+lemma LSeries_ne_zero_of_one_lt_re {N : ℕ} (χ : DirichletCharacter ℂ N) {s : ℂ} (hs : 1 < s.re) :
+    L ↗χ s ≠ 0 :=
+  fun h ↦ by simpa only [h, zero_mul, zero_ne_one] using LSeries.mul_mu_eq_one χ hs
+
+end DirichletCharacter
+
+
+section zeta
+
+/-!
+### The L-series of the constant sequence 1 / the arithmetic function ζ
+
+Both give the same L-series (since the difference in values at zero has no effect;
+see `ArithmeticFunction.LSeries_zeta_eq`), which agrees with the Riemann zeta function
+on `re s > 1`. We state most results in two versions, one for `1` and one for `↗ζ`.
+-/
+
+open LSeries Nat Complex DirichletCharacter
+
+/-- The abscissa of (absolute) convergence of the constant sequence `1` is `1`. -/
+lemma LSeries.abscissaOfAbsConv_one : abscissaOfAbsConv 1 = 1 :=
+  modOne_eq_one (χ := χ₁) ▸ absicssaOfAbsConv_eq_one one_ne_zero χ₁
+
+/-- The `LSeries` of the constant sequence `1` converges at `s` if and only if `re s > 1`. -/
+theorem LSeriesSummable_one_iff {s : ℂ} : LSeriesSummable 1 s ↔ 1 < s.re :=
+  modOne_eq_one (χ := χ₁) ▸ LSeriesSummable_iff one_ne_zero χ₁
+
+
+namespace ArithmeticFunction
+
+/-- The `LSeries` of the arithmetic function `ζ` is the same as the `LSeries` associated
+to the constant sequence `1`. -/
+lemma LSeries_zeta_eq : L ↗ζ = L 1 := by
+  ext s
+  exact (LSeries_congr s const_one_eq_zeta).symm
+
+/-- The `LSeries` associated to the arithmetic function `ζ` converges at `s` if and only if
+`re s > 1`. -/
+theorem LSeriesSummable_zeta_iff {s : ℂ} : LSeriesSummable (ζ ·) s ↔ 1 < s.re :=
+  (LSeriesSummable_congr s const_one_eq_zeta).symm.trans <| LSeriesSummable_one_iff
+#align nat.arithmetic_function.zeta_l_series_summable_iff_one_lt_re ArithmeticFunction.LSeriesSummable_zeta_iff
+-- deprecated 2024-03-29
+@[deprecated] alias zeta_LSeriesSummable_iff_one_lt_re := LSeriesSummable_zeta_iff
+
+/-- The abscissa of (absolute) convergence of the arithmetic function `ζ` is `1`. -/
+lemma abscissaOfAbsConv_zeta : abscissaOfAbsConv ↗ζ = 1 := by
+  rw [abscissaOfAbsConv_congr (g := 1) fun hn ↦ by simp [hn], abscissaOfAbsConv_one]
+
+/-- The L-series of the arithmetic function `ζ` equals the Riemann Zeta Function on its
+domain of convergence `1 < re s`. -/
+lemma LSeries_zeta_eq_riemannZeta {s : ℂ} (hs : 1 < s.re) : L ↗ζ s = riemannZeta s := by
+  simp only [LSeries, natCoe_apply, zeta_apply, cast_ite, cast_zero, cast_one,
+    zeta_eq_tsum_one_div_nat_cpow hs]
+  refine tsum_congr fun n ↦ ?_
+  rcases eq_or_ne n 0 with rfl | hn
+  · simp only [term_zero, cast_zero, ne_eq, ne_zero_of_one_lt_re hs, not_false_eq_true, zero_cpow,
+      div_zero]
+  · simp only [ne_eq, hn, not_false_eq_true, term_of_ne_zero, ↓reduceIte, one_div]
```

Comment:
```suggestion
  rcases eq_or_ne n 0 with hn | hn <;> simp [hn, ne_zero_of_one_lt_re hs]
```

Reply:
This is shorter, but it almost doubles the heartbeats (from 458 to 867).
I am not really sure whether it is better to optimize the proof script (for length and readability) or the proof speed...

Reply:
I've instead optmized the `simp only` calls somewhat.

Reply:
> This is shorter, but it almost doubles the heartbeats (from 458 to 867). I am not really sure whether it is better to optimize the proof script (for length and readability) or the proof speed...

I am never quite sure either how these three priorities should be balanced. (I asked about this on zulip recently but no clear consensus seems to have emerged.) Given the lack of clarity on this, it is clearly up to PR authors to choose their priorities, so please don’t feel bound by my suggestion in any way.